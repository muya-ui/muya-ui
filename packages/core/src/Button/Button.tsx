import React from 'react';
import styled, { css, CSSObject, ThemedStyledProps } from 'styled-components';

import { colorUtils, ITheme } from '@muya-ui/theme-light';

import { withThemeForStyled } from '../utils/withTheme';
import ButtonCore from './ButtonCore';
import { buttonSizeCSS, highlightNeutralColorButtonType, siblingCss } from './mixins';
import { IButtonNode, IButtonProps, IButtonType } from './types';

const ButtonPure = React.forwardRef<IButtonNode, IButtonProps>((props, ref) => {
  const { type, shape, block, groupType, offBorder, plain, selected, ...otherProps } = props;

  return <ButtonCore ref={ref} {...otherProps} />;
});

const buttonInGroupCss = (groupType: IButtonProps['groupType']) => {
  return css`
    ${groupType === 'head' &&
      css`
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      `}

    ${groupType === 'tail' &&
      css`
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
        margin-left: -1px;
      `}
  `;
};

interface IButtonCSSConfig {
  style: CSSObject;
  hoverStyle?: CSSObject;
  clickStyle?: CSSObject;
}
export function mergeCSSConfig(a: IButtonCSSConfig, b: IButtonCSSConfig) {
  return {
    style: {
      ...a.style,
      ...b.style,
    },
    hoverStyle: {
      ...a.hoverStyle,
      ...b.hoverStyle,
    },
    clickStyle: {
      ...a.clickStyle,
      ...b.clickStyle,
    },
  };
}

const buttonBorderCSS = (
  theme: ITheme,
  plain?: boolean,
  groupType?: IButtonProps['groupType'],
  offBorder?: IButtonProps['offBorder'],
  busyOrLoading?: boolean,
) => {
  const { plainIsLight } = theme.components.Button;

  const groupCSSObj: CSSObject = {};
  // 面性的按钮在 ButtonGroup 中需要设置 border
  if ((!plain || plainIsLight) && groupType) {
    if (groupType !== 'tail') {
      groupCSSObj.borderRightWidth = '1px';
      groupCSSObj.borderRightStyle = 'solid';
    }

    if (groupType !== 'head') {
      groupCSSObj.borderLeftWidth = '1px';
      groupCSSObj.borderLeftStyle = 'solid';
    }
  }

  // 在 busy 或者 loading 时整体会有透明度，导致中间的border颜色变深
  // 需要处理一下
  if (busyOrLoading) {
    if (offBorder === 'both' || offBorder === 'right') {
      groupCSSObj.borderRightColor = 'transparent';
    }

    if (offBorder === 'both' || offBorder === 'left') {
      groupCSSObj.borderLeftColor = 'transparent';
    }
  }

  const noBorderGroupCSS = css(groupCSSObj);

  let borderCss;
  if (plain && !plainIsLight) {
    borderCss = css`
      border-style: solid;
      border-width: 1px;
    `;
  } else {
    borderCss = css`
      border-width: 0;
    `;
  }

  return css`
    ${borderCss}
    ${noBorderGroupCSS}
  `;
};

/* eslint-disable complexity */
function getOneTypeOfButton(
  theme: ITheme,
  type: IButtonType,
  interactive: boolean,
  plain?: boolean,
): IButtonCSSConfig {
  const {
    typeColor,
    defaultColor,
    typeBgImage,
    statusOpacity,
    textColor,
    lightTypeColor,
    plainIsLight,
    plainColorHighlight,
    borderColor: typeBorderColor,
    typeBgColor,
  } = theme.components.Button;
  const currentBgImage = typeBgImage[type];
  const currentColor = typeColor[type];
  const currentLightColor = lightTypeColor[type];
  const currentBgColor = typeBgColor[type];
  let currentTypeUseBgImage = !!currentBgImage;

  let borderColor;

  // 当是面性的按钮时，在 ButtonGroup 中是有 border 的
  // 需要设置 borderColor
  if (plain && plainIsLight) {
    borderColor = currentLightColor.hover;
  } else if (plain) {
    borderColor = typeBorderColor[type];
  } else {
    borderColor = currentColor.hover;
  }

  let color: string;
  if (plain) {
    color = currentColor.normal;
  } else if (textColor[type]) {
    color = textColor[type]!;
  } else {
    color = defaultColor;
  }

  let bgColor;
  if (plain && plainIsLight) {
    bgColor = currentLightColor.normal;
  } else if (plain) {
    bgColor = 'transparent';
  } else if (currentBgColor) {
    bgColor = currentBgColor.normal;
  } else {
    bgColor = currentColor.normal;
  }

  let hoverColor;
  let hoverBackgroundImage;
  let hoverBackgroundColor;
  let hoverBorderColor;

  let clickColor;
  let clickBackgroundImage;
  let clickBackgroundColor;
  let clickBorderColor;

  if (interactive) {
    if (plain && !plainIsLight) {
      hoverBackgroundColor = colorUtils.transparentize(statusOpacity.hover, color);
      clickBackgroundColor = colorUtils.transparentize(statusOpacity.click, color);
    } else if (plain) {
      hoverBackgroundColor = currentLightColor.hover;
      clickBackgroundColor = currentLightColor.click;
    } else if (currentTypeUseBgImage) {
      hoverBackgroundImage = currentBgImage!.hover;
      clickBackgroundImage = currentBgImage!.click;
    } else if (currentBgColor) {
      hoverBackgroundColor = currentBgColor.hover;
      clickBackgroundColor = currentBgColor.click;
    } else {
      hoverBackgroundColor = currentColor.hover;
      clickBackgroundColor = currentColor.click;
    }

    if (plain && plainColorHighlight && highlightNeutralColorButtonType(type)) {
      hoverColor = typeColor.primary.hover;
      clickColor = typeColor.primary.click;
      hoverBorderColor = typeColor.primary.hover;
      clickBorderColor = typeColor.primary.click;
    } else {
      hoverColor = color;
      clickColor = color;
    }
  }

  let backgroundImage;
  let backgroundColor;
  // 不是线框的按钮，同时使用背景
  if (!plain && currentTypeUseBgImage) {
    backgroundImage = currentBgImage!.normal;
  } else {
    backgroundColor = bgColor;
  }

  let busyOpacity;
  if (!interactive) {
    busyOpacity = theme.opacity.pattern.disabled;
  }

  return {
    style: {
      color,
      backgroundImage,
      backgroundColor,

      opacity: busyOpacity,
      cursor: interactive ? 'pointer' : 'not-allowed',
      borderColor,
    },
    hoverStyle: {
      color: hoverColor,
      backgroundImage: hoverBackgroundImage,
      backgroundColor: hoverBackgroundColor,

      borderColor: hoverBorderColor || borderColor,
    },
    clickStyle: {
      color: clickColor,
      backgroundImage: clickBackgroundImage,
      backgroundColor: clickBackgroundColor,

      borderColor: clickBorderColor || borderColor,
    },
  };
}

/* eslint-enable complexity */

function getDisabledButtonConfig(theme: ITheme, plain?: boolean): IButtonCSSConfig {
  const { defaultColor, plainIsLight } = theme.components.Button;
  const disabledColor = theme.colors.pattern.text.disabled;

  let color = defaultColor;
  let borderColor;
  let backgroundColor;

  // 如果是线框的
  if (plain && !plainIsLight) {
    color = disabledColor;
    borderColor = disabledColor;
    backgroundColor = 'transparent';
  } else {
    borderColor = defaultColor;
    backgroundColor = disabledColor;
  }

  return {
    style: {
      cursor: 'not-allowed',
      color,
      backgroundColor,
      borderColor,
    },
  };
}

function getButtonConfigInGroup(
  theme: ITheme,
  type: IButtonType,
  selected?: boolean,
): IButtonCSSConfig {
  const { hoverZIndex, selectedZIndex, zIndexOrder } = theme.components.Button;
  const currentZIndex = zIndexOrder[type];
  let activeZIndex;
  if (selected) {
    activeZIndex = selectedZIndex;
  } else {
    activeZIndex = hoverZIndex;
  }

  let normalZIndex;
  if (currentZIndex && selected) {
    normalZIndex = selectedZIndex;
  } else if (currentZIndex) {
    normalZIndex = currentZIndex;
  }
  return {
    style: {
      zIndex: normalZIndex,
    },
    clickStyle: {
      zIndex: activeZIndex,
    },
    hoverStyle: {
      zIndex: activeZIndex,
    },
  };
}

// 状态的样式
function buttonCSS(props: ThemedStyledProps<IButtonProps, ITheme>) {
  const {
    size = 'm',
    type = 'normal',
    plain,
    theme,
    busy,
    disabled,
    loading,
    groupType,
    offBorder,
    selected,
    disableSiblingMargin = false,
  } = props;
  const interactive = !(busy || disabled || loading);
  const baseConfig = disabled
    ? getDisabledButtonConfig(theme, plain)
    : getOneTypeOfButton(theme, type, interactive, plain);
  let config;
  if (groupType) {
    const groupConfig = getButtonConfigInGroup(theme, type, selected);
    config = mergeCSSConfig(baseConfig, groupConfig);
  } else {
    config = baseConfig;
  }

  const normalCss = css(config.style);
  let interactiveCss;
  if (interactive && config.hoverStyle && config.clickStyle) {
    interactiveCss = css`
      &:hover {
        ${css(config.hoverStyle)}
      }
      &:active {
        ${css(config.clickStyle)}
      }
    `;
  }
  const busyOrLoading = busy || loading;

  return css`
    ${normalCss}
    ${interactiveCss}
    ${buttonBorderCSS(theme, plain, groupType, offBorder, busyOrLoading)}
    ${siblingCss(size, theme, !!groupType, disableSiblingMargin)}
  `;
}

const Button = styled(ButtonPure)<IButtonProps>`
  ${props => {
    const { type, selected, groupType, plain, textFine, fontWeight } = props;
    let innerPlain = plain;
    let innerType = type;
    if (!innerType && selected === true) {
      innerType = 'primary';
    } else if (!innerType && selected === false) {
      innerType = 'secondary';
    } else if (!innerType) {
      innerType = 'normal';
    }
    if (innerType === 'normal' && plain === undefined) {
      innerPlain = true;
    }
    let innerFontWeight = fontWeight;
    if (!innerFontWeight && textFine) {
      innerFontWeight = 'lighter';
    }
    const newProps = {
      ...props,
      type: innerType,
      plain: innerPlain,
      fontWeight: innerFontWeight,
    };
    return css`
      user-select: none;
      ${buttonSizeCSS(newProps)}
      ${buttonInGroupCss(groupType)}
      ${buttonCSS(newProps)}
    `;
  }}
`;
(Button as any).__MUYA_BUTTON = true;
export default withThemeForStyled(Button);
