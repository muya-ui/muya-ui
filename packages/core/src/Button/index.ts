export * from './types';
export { default as ButtonGroup } from './ButtonGroup';
export { default as Button } from './Button';
export { default as InlineButton } from './InlineButton';
export { default as MaskButton } from './MaskButton';
export { default } from './Button';
