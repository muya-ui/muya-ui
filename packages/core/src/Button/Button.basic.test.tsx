import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Button from './Button';

test('测试设置按钮 width', () => {
  const tree = renderer
    .create(
      <>
        <Button width={200}>Test</Button>
        <Button width="20%">Test</Button>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('render as link', () => {
  const tree = renderer
    .create(
      <>
        <Button href="test">Test</Button>
        <Button component="a">Test</Button>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 prop shape', () => {
  const tree = renderer
    .create(
      <>
        <Button shape="circle">Test</Button>
        <Button shape="round">Test</Button>
        <Button shape="square">Test</Button>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 prefix suffix', () => {
  const node = <span>aa</span>;
  const tree = renderer
    .create(
      <>
        <Button prefixNode={node}>Test</Button>
        <Button suffixNode={node}>Test</Button>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 prop block', () => {
  const tree = renderer
    .create(
      <>
        <Button block>Test</Button>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('have static perperty for type detecting', () => {
  const wrapper = mount(<Button>按钮</Button>);
  // eslint-disable-next-line
  expect((wrapper.type() as any).__MUYA_BUTTON).toBe(true);
});

test('测试 marginLeft', () => {
  const tree = renderer.create(<Button disableSiblingMargin={true}>Test</Button>).toJSON();
  expect(tree).toMatchSnapshot();
});
