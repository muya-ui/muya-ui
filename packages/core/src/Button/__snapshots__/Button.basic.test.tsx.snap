// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render as link 1`] = `
Array [
  .c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 + .c1 {
  margin-left: 8px;
}

.c4 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

<a
    aria-disabled={false}
    className="c0 c1 c2"
    href="test"
    onClick={[Function]}
    role="button"
  >
    Test
  </a>,
  .c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 + .c1 {
  margin-left: 8px;
}

.c4 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

<a
    aria-disabled={false}
    className="c0 c1 c2"
    onClick={[Function]}
    role="button"
  >
    Test
  </a>,
]
`;

exports[`测试 marginLeft 1`] = `
.c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 + .c1 {
  margin-left: 8px;
}

.c4 + .c1 {
  margin-left: 8px;
}

.c5 + .c1 {
  margin-left: 8px;
}

.c6 + .c1 {
  margin-left: 8px;
}

.c7 + .c1 {
  margin-left: 8px;
}

.c8 + .c1 {
  margin-left: 8px;
}

.c9 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

<button
  className="c0 c1 c2"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Test
</button>
`;

exports[`测试 prefix suffix 1`] = `
Array [
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding-right: 6px;
}

.c3 .c9 {
  line-height: unset;
}

.c4 + .c1 {
  margin-left: 8px;
}

.c5 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

.c6 + .c1 {
  margin-left: 8px;
}

.c7 + .c1 {
  margin-left: 8px;
}

.c8 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span
      className="c3 muya-button-prefix"
    >
      <span>
        aa
      </span>
    </span>
    Test
  </button>,
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding-left: 6px;
}

.c3 .c9 {
  line-height: unset;
}

.c4 + .c1 {
  margin-left: 8px;
}

.c5 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

.c6 + .c1 {
  margin-left: 8px;
}

.c7 + .c1 {
  margin-left: 8px;
}

.c8 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Test
    <span
      className="c3 muya-button-suffix"
    >
      <span>
        aa
      </span>
    </span>
  </button>,
]
`;

exports[`测试 prop block 1`] = `
.c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 + .c1 {
  margin-left: 8px;
}

.c4 + .c1 {
  margin-left: 8px;
}

.c5 + .c1 {
  margin-left: 8px;
}

.c6 + .c1 {
  margin-left: 8px;
}

.c7 + .c1 {
  margin-left: 8px;
}

.c8 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  width: 100%;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

<button
  className="c0 c1 c2"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Test
</button>
`;

exports[`测试 prop shape 1`] = `
Array [
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 + .c1 {
  margin-left: 8px;
}

.c4 + .c1 {
  margin-left: 8px;
}

.c5 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  width: 32px;
  border-radius: 50%;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

.c6 + .c1 {
  margin-left: 8px;
}

.c7 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 + .c1 {
  margin-left: 8px;
}

.c4 + .c1 {
  margin-left: 8px;
}

.c5 + .c1 {
  margin-left: 8px;
}

.c6 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 999rem;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

.c7 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 + .c1 {
  margin-left: 8px;
}

.c4 + .c1 {
  margin-left: 8px;
}

.c5 + .c1 {
  margin-left: 8px;
}

.c6 + .c1 {
  margin-left: 8px;
}

.c7 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  width: 32px;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
]
`;

exports[`测试设置按钮 width 1`] = `
Array [
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  width: 200px;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

.c3 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 + .c1 {
  margin-left: 8px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  width: 20%;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c2:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c2:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c2 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
]
`;
